openfisca_us/data/datasets/cps/cps.py

```diff
@@ -0,0 +1,163 @@
+        """
+
+        # Prepare raw CPS tables
+        year = int(year)
+        if year not in RawCPS.years:
+            logging.info(f"Generating raw CPS for year {year}.")
+            RawCPS.generate(year)
+
+        raw_data = RawCPS.load(year)
+        cps = h5py.File(self.file(year), mode="w")
+
+        person, tax_unit, family, spm_unit, household = [
+            raw_data[entity]
+            for entity in (
+                "person",
+                "tax_unit",
+                "family",
+                "spm_unit",
+                "household",
+            )
+        ]
+
+        add_ID_variables(cps, person, tax_unit, family, spm_unit, household)
+        add_personal_variables(cps, person)
+        add_personal_income_variables(cps, person)
+        add_SPM_variables(cps, spm_unit)
+
+        raw_data.close()
+        cps.close()
+
+
+def add_ID_variables(
+    cps: h5py.File,
+    person: DataFrame,
+    tax_unit: DataFrame,
+    family: DataFrame,
+    spm_unit: DataFrame,
+    household: DataFrame,
+):
+    """Add basic ID and weight variables.
+
+    Args:
+        cps (h5py.File): The CPS dataset file.
+        person (DataFrame): The person table of the ASEC.
+        tax_unit (DataFrame): The tax unit table created from the person table
+            of the ASEC.
+        family (DataFrame): The family table of the ASEC.
+        spm_unit (DataFrame): The SPM unit table created from the person table
+            of the ASEC.
+        household (DataFrame): The household table of the ASEC.
+    """
+    # Add primary and foreign keys
+    cps["person_id"] = person.PH_SEQ * 100 + person.P_SEQ
+    cps["family_id"] = family.FH_SEQ * 10 + family.FFPOS
+    cps["household_id"] = household.H_SEQ
+    cps["person_tax_unit_id"] = person.TAX_ID
+    cps["person_spm_unit_id"] = person.SPM_ID
+    cps["tax_unit_id"] = tax_unit.TAX_ID
+    cps["spm_unit_id"] = spm_unit.SPM_ID
+    cps["person_household_id"] = person.PH_SEQ
+    cps["person_family_id"] = person.PH_SEQ * 10 + person.PF_SEQ
+
+    # Add weights
+    # Weights are multiplied by 100 to avoid decimals
+    cps["person_weight"] = person.A_FNLWGT / 1e2
+    cps["family_weight"] = family.FSUP_WGT / 1e2
+
+    # Tax unit weight is the weight of the containing family.
+    family_weight = Series(
+        cps["family_weight"][...], index=cps["family_id"][...]
+    )
+    person_family_id = cps["person_family_id"][...]
+    persons_family_weight = Series(family_weight[person_family_id])
+    cps["tax_unit_weight"] = persons_family_weight.groupby(
+        cps["person_tax_unit_id"][...]
+    ).first()
+
+    cps["spm_unit_weight"] = spm_unit.SPM_WEIGHT / 1e2
+
+    cps["household_weight"] = household.HSUP_WGT / 1e2
+
+
+def add_personal_variables(cps: h5py.File, person: DataFrame):
+    """Add personal demographic variables.
+
+    Args:
+        cps (h5py.File): The CPS dataset file.
+        person (DataFrame): The CPS person table.
+    """
+    cps["age"] = np.where(
```

Comment:
Please add a comment here on what this is doing and why

Reply:
Added, does this make sense?